service/src/main/java/bio/terra/externalcreds/services/PassportService.java

```diff
@@ -0,0 +1,73 @@
+package bio.terra.externalcreds.services;
+
+import bio.terra.common.db.ReadTransaction;
+import bio.terra.externalcreds.ExternalCredsException;
+import bio.terra.externalcreds.dataAccess.GA4GHPassportDAO;
+import bio.terra.externalcreds.models.GA4GHPassport;
+import bio.terra.externalcreds.models.LinkedAccount;
+import java.sql.Timestamp;
+import java.time.Instant;
+import java.util.Optional;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.security.oauth2.core.OAuth2RefreshToken;
+import org.springframework.security.oauth2.core.endpoint.OAuth2AccessTokenResponse;
+import org.springframework.stereotype.Service;
+
+@Service
+@Slf4j
+public class PassportService {
+
+  private final GA4GHPassportDAO passportDAO;
+  private final OAuth2Service oAuth2Service;
+  private final LinkedAccountService linkedAccountService;
+  private final ProviderClientCache providerClientCache;
+
+  public PassportService(
+      GA4GHPassportDAO passportDAO,
+      OAuth2Service oAuth2Service,
+      LinkedAccountService linkedAccountService,
+      ProviderClientCache providerClientCache) {
+    this.passportDAO = passportDAO;
+    this.oAuth2Service = oAuth2Service;
+    this.linkedAccountService = linkedAccountService;
+    this.providerClientCache = providerClientCache;
+  }
+
+  @ReadTransaction
+  public Optional<GA4GHPassport> getGA4GHPassport(String userId, String providerId) {
+    return passportDAO.getPassport(userId, providerId);
+  }
+
+  public void authAndRefreshPassport(LinkedAccount linkedAccount) {
+    if (!linkedAccount.getExpires().after(Timestamp.from(Instant.now()))) {
+      passportDAO.deletePassport(linkedAccount.getId().get());
+    } else {
+      var clientRegistration = providerClientCache.getProviderClient(linkedAccount.getProviderId());
+      Optional<OAuth2AccessTokenResponse> accessTokenResponse;
+      try {
+        accessTokenResponse =
+            Optional.ofNullable(
+                oAuth2Service.authorizeWithRefreshToken(
+                    clientRegistration,
+                    new OAuth2RefreshToken(linkedAccount.getRefreshToken(), null)));
+      } catch (Exception e) {
+        // An IllegalArgumentException is thrown when the issuer can not be reached
+        // An OAuth2AuthorizationException is thrown when the provider returns a non-200 response
+        // TODO: decide how to handle different exceptions differently
+        //   (upon failure to auth, consider deleting the linked account?)
+        log.error("Encountered an error while getting a new refresh token: {}", e.getMessage());
```

Comment:
this logs just the exception message and looses all other information like exception class and stack trace. You should almost never do this. Instead
```
log.error("Encountered an error while getting a new refresh token", e);
```